aten/src/ATen/SparseCsrTensorImpl.cpp

```diff
@@ -0,0 +1,153 @@
+#include <ATen/ATen.h>
+#include <ATen/InitialTensorOptions.h>
+#include <ATen/SparseCsrTensorImpl.h>
+#include <ATen/SparseTensorImpl.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+
+namespace at {
+namespace {
+DeviceType SparseCsrTensorSetToDeviceType(DispatchKeySet key_set) {
+  if (key_set.has(DispatchKey::SparseCsrCPU)) {
+    return kCPU;
+  } else if (key_set.has(DispatchKey::SparseCsrCUDA)) {
+    return kCUDA;
+  } else {
+    AT_ERROR(
+        "Cannot construct SparseTensor with non-sparse tensor type ID ",
+        key_set);
+  }
+}
+} // namespace
+
+SparseCsrTensorImpl::SparseCsrTensorImpl(
+    at::DispatchKeySet key_set,
+    const caffe2::TypeMeta& data_type)
+    : SparseCsrTensorImpl(
+          key_set,
+          data_type,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(ScalarType::Int)) // crow_indices
+          ,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(ScalarType::Int)) // col_indices
+          ,
+          at::empty(
+              {0},
+              at::initialTensorOptions()
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(data_type)) // values
+      ) {}
+
+SparseCsrTensorImpl::SparseCsrTensorImpl(
+    at::DispatchKeySet key_set,
+    const caffe2::TypeMeta& data_type,
+    at::Tensor crow_indices,
+    at::Tensor col_indices,
+    at::Tensor values)
+    : TensorImpl(key_set, data_type, values.device()),
+      crow_indices_(std::move(crow_indices)),
+      col_indices_(std::move(col_indices)),
+      values_(std::move(values)) {}
```

Comment:
dtypes and devices of tensors are not validated here; this is OK but need to make sure it is validated at all call sites

Reply:
What is less OK is that this is inconsistent with `set_member_tensors`, which DOES do validation on the passed in tensors...

Reply:
Given that we don't have a solid use case of setting up a CSR object and then assigning it (unlike some places in COO code), I will stick to your suggestion and make this constructor the single point of entry for setting up and checking the members of the CSR tensor.